### ISSUE TITLE
Heatmap content is flipped vertically

### ISSUE THREAD

**Issue by vtozarks** [ScottPlot/ScottPlot#1864]
## Bug Report

**Issue:** (Bug description goes here)
When plotting data in the Heatmap the data is flipped vertically.

**Reproducing:** (What do others have to do to reproduce the bug?)
You can see this even in the plots in the Cookbook Heatmap examples.  The "coldest" point should be at (0,0) but it is on the max Y row  and the "hottest" point should be on the max Y row but it is on the min Y row in the plot.

It appears this is happening in the "Update" function.  The 2D data is flattened into a 1D array.  Later in that routine the Rectangle is defined with the upper left corner set at 0,0.  I believe the flattened data is then used to populate that Rectangle from that upper left corner down.  In playing with this is I added the line of code shown below at the end of the Update function and I get the expected results.  I am not sure if there is a more elegant/efficient way to solve this.

```cs
// a code sample may improve communication
BmpHeatmap.RotateFlip(RotateFlipType.RotateNoneFlipY);
```

### System Details
* **ScottPlot Version:** 4.1.123?
* **Operating System:** (Windows 10.)
* **Application Type:** (console)
* **.NET Version:** (.NET 6.0)

**Comment by bclehmann** [comment]
Hi, if I understand you correctly, you expect this code:
```cs
double[,] data2D = { { 1, 2, 3 },
         { 4, 5, 6 } };

plt.AddHeatmap(data2D, lockScales: false);
```

To produce this image (ignore that the axes are also flipped):
![image](https://user-images.githubusercontent.com/8635304/170890490-57500729-c259-456f-b699-af6becad685a.png)

But in reality, it produces this one:
![image](https://user-images.githubusercontent.com/8635304/170890506-ef44d581-92d4-4d25-a3e0-e844a7a37416.png)

<hr/>

This is actually the expected behaviour, unintuitive as it may seem. Heatmaps are drawn with starting from the top-left corner rather than the bottom-left. The main reason ScottPlot went in this direction is for consistency with (many but not all) other plotting libraries. And other plotting libraries went this direction because heatmaps are often used to show matrices (including images) that are drawn from top-to-bottom. This is also (usually) the more convenient choice when working with real-world data from sensors.

However, I recognize it is inconvenient when creating a heatmap from a mathematical function, because things are flipped in the x-axis. This is actually how I use heatmaps most often, so I'm surprised I never added an option to flip them.

**Comment by bclehmann** [comment]
FWIW, [`Graphics.ScaleTransform`](https://docs.microsoft.com/en-us/dotnet/api/system.drawing.graphics.scaletransform?view=dotnet-plat-ext-6.0) is probably the easiest (and most efficient) way to get this working, assuming it supports a scale of `-1`

I'm leaning towards an API that just has `bool FlipVertically { get; set; }` and potentially `bool FlipHorizontally { get; set; }`, as that seems simpler than an enum.

**Comment by bclehmann** [comment]
There is potential funny business if we ever decide to support rotation, as a rotation by angle $\theta$ followed by a flip is not the same as a flip followed by a rotation. I don't know what other libraries do (I suspect most plotting libraries don't allow heatmap rotation), but there may be a precedent somewhere we can mimic.

**Comment by vtozarks** [comment]
Rotate would be interesting/useful at times but as you said that can lead to a lot of confusion.  I think a simple boolean flip makes sense.  

I guess if top-to-bottom is the norm for heatmaps (I was not aware of that) then at least the axis values should reflect that plotting decision and perhaps it should be noted in the examples.  This is what was fundamentally confused me - the coordinates on the plot did not line up with the coordinates of my data.  With or without the new flip boolean I think it is important that the coordinates accurately indicate how the data is being presented to prevent someone only looking at the plot from making a bad conclusion.

**Comment by bclehmann** [comment]
I looked to see what other libraries do, and while axes from the top is more common, it's not universal the board.

It looks like matplotlib, plotly, and seaborn both have the y-axis start from the top, where gnuplot, and MATLAB have it start from the bottom. All of these libraries draw from the top left. I also looked at OxyPlot and I'm unsure where it starts drawing, but the axis starts from the bottom.

If I were to rewrite it today I would probably flip the axis, but I wouldn't do that now because I don't think the current behaviour is wrong so much as it does the less popular (and less intuitive) of the two common options. Backwards compatibility, in my mind, is more important. Especially as we do document how to adjust the axes (although it perhaps could be better).

**Comment by vtozarks** [comment]
I agree - you would not want to make 'flip' the new default and break existing users so I think a boolean to allow it makes the most sense.  

I would however still think that the current behavior should adjust the axis tick labels so they line up with how the data is being presented (i.e. 0,0 should be in the upper left).  Do the other tools doing top-to-bottom also have (0,0) shown in the lower left?  I don't have any existing test cases.

**Comment by vtozarks** [comment]
Here is an example OxyPlot I got from a colleague (plot for data at (+/-1, +/-1).  In a discussion with a couple engineers in my office several flagged that the other tools are centering each square on the raw data coordinate and raised some concern that the ScotPlott example was effectively building in a (0.5, 0.5) offset that is not present in the raw data.  All agreed that if coordinates are shown on the axis then they should align with the raw data to prevent inaccurate conclusions from the plot no matter which way the data is plotted (starting at upper left or lower left).

![OxyPlotHeatMap](https://user-images.githubusercontent.com/24811236/171258855-64a48e22-7e97-498e-807a-91fc0247c038.png)

**Comment by bclehmann** [comment]
Aye, I think it would be valuable to have the axis lines run through the pixels rather than through their top-left corner, but I don't think we can fix that without a major version bump for backwards-compatibility reasons. If we were to fix it, users who use `plt.SetAxisLimits` would have their heatmap shifted.

@swharden Perhaps there ought to be a section in #1028 for changes we'd like to make but cannot until ScottPlot 5?

**Comment by swharden** [comment]
Thanks for reaching out @vtozarks! The fix @bclehmann offered in #1866 looks great and I'm merging it in now.

Regarding simpler ways to do this in ScottPlot5, I don't even think we need to note it in #1028. Now that the heatmap module has been running in the wild for a long time I think we have a better idea than when we started about what parts people find intuitive vs. unintuitive/surprising, and could probably design a new API around the 99% use case, and not worry about breaking changes. The time isn't quite right to dive into that yet, but I look forward to hopefully working on it soon!
